### ISSUE TITLE
Do we have a guideline about setup a thumbor cluster on Cloud?

### ISSUE THREAD

**Issue by Tideorz** [thumbor/thumbor#1144]
Thumbor is a very powerful image service,  and I think it should have a way to deploy thumbor cluster on 
cloud, do we have any best practice for this? Best is about the thumbor on Kubernetes, other cloud platform's knowledge is also fine.
I'm curious about some questions:
1. how to monitor multiple thumbor instances in cloud? 
2. how many thumbor threads should be running in one container? the number of CPUs of one container?

Cheers

**Comment by dhardy92** [comment]
"Servers ? Where we're going we don't need... servers."  Doc Brown

Amazon AWS documentation wrote an Implementation guide to run Thumbor in serverless architecture using AWS lambda and AWS API Gateway https://docs.aws.amazon.com/fr_fr/solutions/latest/serverless-image-handler/welcome.html

This need a read at least.

**Comment by dhardy92** [comment]
And you think well.
I may answer each points as I implemented it successfully for some real size tests.
I ran 2 Thumbor EC2 instances managed with puppet for a part of my image traffic. And at the end of 2018 I give a naive try on this lambda based implementation (still running) so here is my 1 month experience feed back with Tumbor AWS Lambda 

* simplicity: as long as CloudFormation template is provided, the CF, API Gateway and Lambda stack complexity is delegated (and it helped me a lot to understand how those things interacts BTW)

* cost: YMMV is very true. Lambda fit really well with usage and seasonality of the traffic. In my case : I run 2 EC2 c4.xlarge for thumbor Debian based instances (409.54$/month) no RI or Spot for legacy archi.
           Using lambda it vary a lot days to days by I have 1 month of usage (yes christmas holidays in as bias) with a total cost of 449.29$ for the same period. 

* pace: True too. I fork it and made some changes. In the previous versions they even broke the url signature usage putting secret keys in URL (fixed since). I did some change in the code to bump thumbor and tc-aws versions. It work well. I did other changes for my own needs (default image on 404 for some Ids) and also a PR to use KMS to encrypt secret https://github.com/awslabs/serverless-image-handler/pull/77/files . Wait and see for now.

* features: I do not use much features in my case. But when building archive I encounter some   problems on optimizers (not dig so much on it for now) all python things seams OK. Config is customizable via ENV variables.

* speed: cold start is a thing . In my use case first call are about 600ms then 125ms averages. Cloudfront mitigates it in a way. I have ~1 error / 3hours (>10s) otherwise max is ~2s (sadly Cloudwatch is not showing me p90 or so. Can't be more precise). Invocation/min are between 2k at night to 20k on day pic.

* caching: disk caching is still possible but very volatile I suppose. Maybe redis/s3 caching make more sense here.

This stack is very educative at least.

PS: I have some graphs to illustrate but github "can't process that file" ...

**Comment by kkopachev** [comment]
Lambda could certainly be an option.
For our main site we tried running thumbor@lambda when it first got available. For our workload, it was unusable, even if patched manually to cover our cases. 
Price was comparable or more expensive to on-demand EC2 instances. We do gifv conversion on thumbor and for some big gifs we hit the timeout (don't remember where exactly). After all we got a bunch of errors on cloudfront for images, but there were no indication of throttling on lambda side. 
Since then, we moved to spot instances, which made the thing even cheaper.
So our setup is an ECS cluster (container scaling) of spot instances (bunch of different types), managed by spot fleet (instances scaling).

We do use thumbor@lambda for few small projects, though.

tldr: YMMV

**Comment by heynemann** [comment]
This is possibly THE best thread in our repo. Thanks everyone for very valuable inputs.

**Comment by gingerlime** [comment]
Very interesting pointer about spot instances @kkopachev !

Yes, I'm super curious about the different use cases, deployments and optimizations, and it's great to get some of those insights, especially from those who tried a few options. I hope to collect some recipes and best practice over time, so this shared knowledge can benefits others.

**Comment by ElMoselYEE** [comment]
We've been looking at running the serverless Lambda version as well, so glad others are considering similar paths. 

I have some thoughts and also some questions:

* we forked the repo as we needed to customize the error handling and metrics dispatching. Also noticed that timeouts always caused the next request to 500 because the server is busy, so had to patch that.
* regarding simplicity: building the lambda is complex and I worry it's fragile and may one day just stop working, but running lambdas is way less maintenance on our side as we don't have to worry about patching OS security updates nor monitoring CPU / memory usage. 
* pergormance: we have enough traffic, about 25 mil req/day, where we don't expect the cold start situation to be that impactful, but time will tell. 
* cacheing: we plan on using CloudFront to cache all rendered images for the lifetime of the image: about 90 days. All of our images are already in S3. I'm curious where you feel this falls short of capabilities a server would have. In what way would enabling thumbor-side caching improve response times?

**Comment by gingerlime** [comment]
Yes, scaling without worrying about instances etc is a very appealing value proposition. I'm generally a big fan of Lambda (I even created an [A/B testing framework](https://github.com/Alephbet/gimel) running on it).

But given how many dependencies and libraries thumbor typically needs, this feels like really stretching Lambda. I think it excels with simple functions, but as they grow in complexity, things get much more tricky. That said, the guys at AWS have put good effort into it, and it's by and large a black box that "just works" at this stage. So maybe my fear is largely irrational, or maybe I'm just finding reasons to justify my personal bias :)

To answer your question on caching from my personal experience, primarily with Cloudflare. They have dozens or hundreds of edges, so if your traffic is distributed widely, people will hit different edges, and those edges won't all have the cached image. Combine that with some heavy transformations with thumbor (e.g. smart cropping, filters, watermarks etc), it can get slow and expensive. Mind you, this is more of a cost issue with Lambda or a scaling/capacity issue if you're running your own servers (unless you worry about few users getting slow experience occasionally too). 

I found that nginx-proxy with local file result storage solves this problem pretty well, in addition to CDN caching (not replacing it). That's why I included it on the MinimalCompact repo, and there are some examples of how to combine the two.

Of course, the more servers / local caches you use, you increase your cache miss rate. This is more manageable with a relatively small amount of server instances though. I'm not entirely sure about the economics for redis, S3 etc...

**Comment by kkopachev** [comment]
Lambda has limit of max response body hard set to 6MB. So if you're trying to serve image bigger than that, then lambda is not an option. 
We just hit that on one of smaller projects when trying to serve decent size gif.

**Comment by jyrkih** [comment]
We forked it, too. I'd say Lambda is ok in "99%" of cases. And very economical. The problem is that 1% which in our case typically means that huge png/gif file processing fails due to either 1) out of memory (3008MB), or 2) out of execution time (<30s). Also 6MB response size is an issue, we go around that by setting max_bytes filter accordingly. But due to these limitations, I think Fargate would be better option.
